mlir/lib/Quantum/Utils/QuantumSplitting.cpp

```diff
@@ -217,6 +218,24 @@ void AugmentedCircuitGenerator::generate(Region &region, OpBuilder &builder)
         else if (auto whileOp = dyn_cast<scf::WhileOp>(&op)) {
             visitOperation(whileOp, builder);
         }
+        else if (auto callOp = dyn_cast<func::CallOp>(op)) {
+            auto results = callOp.getResultTypes();
+
+            bool multiReturns = results.size() > 1;
+            bool quantum;
+            for (Type res : results) {
+                quantum = isa<QuregType>(res);
+                break;
+            }
```

Comment:
nitpick: If you can remove this loop, I would prefer it :)

Reply:
Just out of curiosity, what would be an alternative? :)

Reply:
The results variable is a result type range. It probably supports front. It is already guaranteed to be bigger than one so calling front would be safe.

bool quantum = isa<QuregType>(results.front())

Reply:
Ah right, with the assertion below that makes sense 👍

Reply:
Actually, the original code is a bug. If results is empty, then quantum is not set.

Reply:
The qreg could be anywhere in the return, so I used C++ any_of. That should return false if it is empty.

```
            bool quantum = std::any_of(results.begin(), results.end(),
                                       [](const auto &value) { return isa<QuregType>(value); });
```

Reply:
Can it really be anywhere? I thought by definition it would always be the first one?

Reply:
if you have a QNode that returns: float, expval, float that would cover this case I think. Also potentially some MLIR transforms could create a mix between qreg and other types.

Reply:
Can you provide a concrete example? Like actually code it and run it through the compilation pipeline and produce a qreg that is not in the first position? I mean, what you wrote is indeed more general, but I find some value in knowing that the qreg will always be returned first.

Reply:
There is no reorderding for example:
```
dev = qml.device("lightning.qubit", wires=2)
@qjit(keep_intermediate=True)
@qml.qnode(device=dev)
def workflow1(a, b):
    qml.RX(a, wires=0)
    return (a, b, qml.probs(), a, b)

a = jnp.array(2.2)
b = jnp.array(3.3)
result = workflow1(a, b)
```
```
  func.func private @workflow1(%arg0: tensor<f64>, %arg1: tensor<f64>) -> (tensor<f64>, tensor<f64>, tensor<4xf64>, tensor<f64>, tensor<f64>) attributes {diff_method = "parameter-shift", llvm.linkage = #llvm.linkage<internal>, qnode} {
    quantum.device ["kwargs", "{'shots': 0}"]
    quantum.device ["backend", "lightning.qubit"]
    %0 = quantum.alloc( 2) : !quantum.reg
    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
    %extracted = tensor.extract %arg0[] : tensor<f64>
    %2 = quantum.custom "RX"(%extracted) %1 : !quantum.bit
    %3 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
    %4 = quantum.compbasis %2, %3 : !quantum.obs
    %5 = quantum.probs %4 : tensor<4xf64>
    %6 = quantum.insert %0[ 0], %2 : !quantum.reg, !quantum.bit
    quantum.dealloc %6 : !quantum.reg
    return %arg0, %arg1, %5, %arg0, %arg1 : tensor<f64>, tensor<f64>, tensor<4xf64>, tensor<f64>, tensor<f64>
  }
```
if you want to work on this function (quantum transform in MLIR) you need to replace %5 by %6 and then you have the QReg in the middle of the return.

Reply:
Nothing wrong with checking this, but in general I think we should avoid relying on such assumptions when the cost isn't too great, especially when working across our main components (e.g. frontend vs. compiler).

Reply:
Mmm... I'm a bit confused. But I'll just yield to you. I think checking all of them is acceptable. What I was asking was whether we would at any point return the qreg not in the first position. What you showed here doesn't really prove that it can be at any other position. But it is a minimal check and more general, so ok.

Reply:
> Nothing wrong with checking this, but in general I think we should avoid relying on such assumptions when the cost isn't too great, especially when working across our main components (e.g. frontend vs. compiler).

I am not sure but I believe you are arguing against my point. I yielded, but I think having strong invariants (which include only returning qreg on the first position) is a better way to code because it would allow us to verify that something went wrong in a transformation. I always prefer strong invariants, maintaining invariants, and checking for invariants. But I do agree that sometimes they get in the way.

Reply:
I think invariants are great! But the positional nature of a value seems rather arbitrary and brittle in a general purpose function 🤔 If this was a custom operation I might be inclined to agree with you. Maybe that's argument for making a quantum function/call operation?

Reply:
I would support it!